File: l33tcode/shortest-path-in-binary-matrix.py
import heapq
from typing import List, Tuple, Iterator, Union


class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0]) if grid else 0

        def neighbors(row: int, col: int) -> Iterator[Tuple[int, int]]:
            for neigh_row, neigh_col in (
                (row - 1, col),
                (row, col - 1),
                (row + 1, col),
                (row, col + 1),
                (row - 1, col - 1),
                (row + 1, col + 1),
                (row - 1, col + 1),
                (row + 1, col - 1),
            ):
                if (
                    0 <= neigh_row < rows
                    and 0 <= neigh_col < cols
                    and grid[neigh_row][neigh_col] == 0
                ):
                    yield neigh_row, neigh_col

        def distance_squared(row1: int, col1: int, row2: int, col2: int) -> int:
            return (row1 - row2) ** 2 + (col1 - col2) ** 2

        heap: List[Tuple[int, int, int, int]] = []
        distances: List[List[Union[float, int]]] = [
            [float("+inf")] * cols for _ in range(rows)
        ]

        if grid[0][0] == 0:
            heapq.heappush(heap, (0, 0, 0, distance_squared(rows - 1, cols - 1, 0, 0)))

        while heap:
            row, col, distance, _ = heapq.heappop(heap)

            if (row, col) == (rows - 1, cols - 1):
                return distance + 1

            if distance >= distances[row][col]:
                continue

            distances[row][col] = distance

            for neigh_row, neigh_col in neighbors(row, col):
                heapq.heappush(
                    heap,
                    (
                        neigh_row,
                        neigh_col,
                        distance + 1,
                        distance_squared(rows - 1, cols - 1, row, col),
                    ),
                )

        return -1

File: l33tcode_rs/src/shortest_path_in_binary_matrix.rs
#[allow(dead_code)]
struct Solution;

use std::cmp::Reverse;
use std::collections::{BinaryHeap, HashSet};

fn neighbours(grid: &[Vec<i32>], row: usize, col: usize) -> Vec<(usize, usize)> {
    let (rows, cols) = (grid.len(), grid[0].len());

    let mut res = Vec::new();

    for (neigh_row_opt, neigh_col_opt) in [
        (Some(row + 1), Some(col)),
        (Some(row), Some(col + 1)),
        (row.checked_sub(1), Some(col)),
        (Some(row), col.checked_sub(1)),
        (row.checked_sub(1), Some(col + 1)),
        (Some(row + 1), Some(col + 1)),
        (row.checked_add(1), col.checked_sub(1)),
        (row.checked_sub(1), col.checked_sub(1)),
    ] {
        if let (Some(neigh_row), Some(neigh_col)) = (neigh_row_opt, neigh_col_opt) {
            if neigh_row < rows && neigh_col < cols && grid[neigh_row][neigh_col] == 0 {
                res.push((neigh_row, neigh_col));
            }
        }
    }

    res
}

impl Solution {
    #[allow(dead_code, clippy::needless_pass_by_value)]
    pub fn shortest_path_binary_matrix(grid: Vec<Vec<i32>>) -> i32 {
        if grid[0][0] == 1 {
            return -1;
        }
        let mut heap: BinaryHeap<(Reverse<usize>, usize, usize)> = BinaryHeap::new();
        let mut visited = HashSet::new();
        let (rows, cols) = (grid.len(), grid[0].len());

        heap.push((Reverse(1), 0, 0));
        visited.insert((0, 0));

        while !heap.is_empty() {
            let (Reverse(distance), row, col) = heap.pop().unwrap();

            if (row, col) == (rows - 1, cols - 1) {
                return distance.try_into().unwrap();
            }

            for (neigh_row, neigh_col) in neighbours(&grid, row, col) {
                if visited.contains(&(neigh_row, neigh_col)) {
                    continue;
                }
                heap.push((Reverse(distance + 1), neigh_row, neigh_col));
                visited.insert((neigh_row, neigh_col));
            }
        }

        -1
    }
}
